Email course- Anchor positioning (#826)
oddbird/oddleventy: content/learn/courses/anchor-positioning.md

---
eleventyExcludeFromCollections: true
title: Create Performant Layouts and Resilient Dropdowns

> sub: This CSS anchor positioning coursegc teaches you how

Author: That was a very generous interpretation :-D. 100% a typo that was meant to go in my terminal.

      but I love what can be done with it.
      Shout out to @OddBird for the great polyfill too
      so I could get it working in Firefox.
    slug: shout-out
    name: Luke Hansford
    title: Engineering Manager, Fishbrain
    venue: "@lukehansford@mastodon.social"
    face: luke-hansford-bg.jpg
  - text: |
      James helped me get my head around anchor positioning quickly. After he explained it to me I was able to put anchor positioning to work in a project immediately, in a way that I would have previously solved with lots of custom JavaScript and headaches. If you want to learn anchor positioning in a fast and practical way, there's no one better than James to teach you.
    name: Jason Lengstorf
    venue: Code.TV
    title: Learn with Jason
    face: jason-lengstorf.webp
    slug: head-around
    url: https://codetv.dev/series/learn-with-jason/
included:
  - title: Immediate access
    text: To lesson 1
    icon: icons/mail-open
  - title: Offer for additional
    text: Free frontend tips & tricks
    icon: icons/tools
  - title: Easily unsubscribe
    text: Whenever you want
    icon: icons/interact
  - title: We won’t share
    text: Your email address

>     icon: icons/eye-crossed

Author: I rewrote "Offer for additional" to-
 
 - title: Easily upgrade for more free tips & tricks
    text: After the course, choose to receive updates from OddBird

### Full support now

Not all your users have a browser that supports CSS anchor positioning yet.
I'll show you how to use OddBird's anchor position polyfill to support all your users.

{% endcall %}

{% call layout.block('column') %}

### Advanced use cases

The ability to declaratively reference the size and position
of another element unlocks all sorts of creative possibilities.

{% endcall %}

{% endcall %}

{% endcall %}

## Anchor Positioning Demo

{{ embed.codepen(
  id='wBvEavV',
  title='Anchor Positioning Demo',
  user='jamessw',
  height=300
) }}

## 9 Weekly Emails

{% call layout.grid(
  'narrow-columns',
  attrs={'style': '--column: page'}
) %}

{% call layout.block('column') %}

- Placement with `position-area`
- Using a position with `anchor()`
- Using a size with `anchor-size()`
- Reusing anchor declarations
- Positioning popovers

{% endcall %}

{% call layout.block('column') %}

- Troubleshooting anchor positioning
- Scroll visibility
- Handling element overflow
- Use it now with a polyfill

{% endcall %}

{% endcall %}

{% call layout.block() %}

{{ utility.main_action(
  content='Sign up, get first email',
  url='#sign-up'
) }}

{% endcall %}

> <div style="margin-bottom: 1rem">

Author: @SondraE felt it was too close. Ideally, there would be a full-width block with a background, with the content still just the normal body width. With the `has-angles` block, the text was too wide.

overflow, anchoring with the Shadow DOM, positioning popovers, using Dev Tools
to troubleshoot anchors, pseudo elements that can be anchors, and new
possibilities for innovative layouts.

</div>

## About James

{% set bird = collections.birds | authorPage('james') %}

{% call birds.card(
  bird=bird,
  extended=extended
) %}

{% endcall %}

{{ utility.main_action(
  content='Subscribe to the free course',
  url='#sign-up'
) }}

{{ quotes.grid(press) }}

<a id="sign-up"></a>
## Create Performant Layouts and Resilient Dropdowns

{% call layout.grid(
  'narrow-columns',
  attrs={'style': '--column: page'}
) %}

{% call layout.block('column') %}

{{ embed.icon_block(
  data=included,
  grid='fit'
) }}

{% endcall %}

{% call layout.block('column') %}

### Sign up for email course
{{ oddnews.oddnews_signup(
  heading=false,
  tag="course-anchor-positioning",
  class="headerless",
  buttonText="Subscribe"
) }}

{% endcall %}

{% endcall %}

## Why it's time to learn CSS anchor positioning

Anchor positioning is currently not available in all browsers,
and it will be 2 to 3 years before you can safely assume all your users have it.

> But now is the time to add this new tool to your toolkit. Here's why.

Author: I think it's fine without.

The ability to describe anchor positioning in a more declarative way has changed
how I think about layout. It’s unlocked new possibilities and made for better
user experiences. And with the polyfill, all your users can have the same
experience without waiting for browsers to catch up.

> ## Don't add another dependency

Author: I mean, yes... but... 

I added the word "forever", I think that differentiates.

  url: "#sign-up"
image:
  src: blog/2024/anchor1.jpg
  alt: A rusty anchor hanging with the sea in the background.
press:
  - text: |
      Just tried out the anchor positioning API in CSS for the first time.
      It's a little more complex than I initially thought it'd be,
      but I love what can be done with it.
      Shout out to @OddBird for the great polyfill too
      so I could get it working in Firefox.
    slug: shout-out
    name: Luke Hansford
    venue: "@lukehansford@mastodon.social"
    url: https://mastodon.social/@lukehansford
    face: luke-hansford-bg.jpg
  - text: |
      James helped me get my head around anchor positioning quickly. After he
      explained it to me I was able to put anchor positioning to work in a
      project immediately, in a way that I would have previously solved with
      lots of custom JavaScript and headaches. If you want to learn anchor
      positioning in a fast and practical way, there's no one better than James
      to teach you.
    name: Jason Lengstorf
    venue: Learn With Jason
    face: jason-lengstorf.webp
    slug: head-around
    url: https://codetv.dev/series/learn-with-jason/
included:
  - title: Immediate access

>     text: You will receive lesson 1 within minutes of subscribing

Author: Not quite- I think that's a blocker for publishing this (and allowing signups in the first place).

    icon: icons/mail-open
  - title: Easily unsubscribe
    text: Take in as much of the course as you'd like
    icon: icons/interact
  - title: Privacy matters to us
    text: We won’t share your email address
    icon: icons/eye-crossed

>   - title: Easily upgrade for more free tips & tricks

Author: Yup! I like this better.

    text: After the course, choose to receive updates from OddBird
    icon: icons/tools
summary: |
  Make your dropdowns more performant and discover new layout possibilities with
  quicker turnarounds using CSS anchor positioning. JavaScript positioning
  libraries require updates and maintenance, but CSS anchor positioning is part
  of the browser spec, and will stand the test of time.
---

{% import "layout.macros.njk" as layout %}
{% import "quotes.macros.njk" as quotes %}
{% import "birds.macros.njk" as birds %}
{% import 'embed.macros.njk' as embed %}
{% import 'utility.macros.njk' as utility %}
{% import 'oddnews.macros.njk' as oddnews %}

{% call layout.block(
  name='full',
  attrs={'class': 'has-angles'}
) %}

## Take this free course to get...

{% call layout.grid('narrow-columns') %}

{% call layout.block('column') %}

### Troubleshooting tips

Most of the time, CSS anchor positioning just works.
But when it doesn't, it can be a real pain.
I'll share the common issues I've seen and how to avoid them.

{% endcall %}

{% call layout.block('column') %}

### Fewer dropdown bugs

> What happens to your dropdowns that are close to the edges,

Reviewer: ```suggestion
What happens to dropdowns that are close to viewport edges,
```

or as the user scrolls?
CSS anchor positioning offers a robust feature set to handle
these literal edge cases.

{% endcall %}

{% call layout.block('column') %}

### Fast page performance

> JavaScript positioning libraries require a lot render-blocking

Reviewer: ```suggestion
JavaScript positioning libraries require a lot of render-blocking
```

---

> eleventyExcludeFromCollections: true

Reviewer: Remove this before merge, if we're ready to launch? Otherwise remove the link from the CSS Anchor Positioning talk page.

{% call layout.block('column') %}

- Troubleshooting anchor positioning
- Scroll visibility
- Handling element overflow
- Use it now with a polyfill

{% endcall %}

{% endcall %}

{% call layout.block() %}

{{ utility.main_action(
  content='Sign up to get your first email',
  url='#sign-up'
) }}

{% endcall %}

<div style="margin-bottom: 1rem">

## What this email course teaches

You will learn about anchor positioning, scoping your anchors, handling
overflow, anchoring with the Shadow DOM, positioning popovers, using Dev Tools
to troubleshoot anchors, pseudo elements that can be anchors, and new
possibilities for innovative layouts.

</div>

## About James

{% set bird = collections.birds | authorPage('james') %}

{{ birds.card(bird=bird, extended=extended) }}

{{ utility.main_action(
  content='Subscribe to the free course',
  url='#sign-up'
) }}

{{ quotes.grid(press) }}

## <a id="sign-up">Create performant layouts and resilient dropdowns</a>

{% call layout.grid(
  'narrow-columns',
  attrs={'style': '--column: page'}
) %}

{% call layout.block('column') %}

{{ embed.icon_block(
  data=included,
  grid='fit'
) }}

{% endcall %}

{% call layout.block('column') %}

> ### Sign up for email course

Reviewer 2: Instead of having this heading outside of the form, should we change the oddnews_signup macro to allow alternative headings/subheadings? I think then the heading would appear to be more associated with the form.

Reviewer 3: I do think it would be nice to switch to the oddnews_signup style here for the reasons you list, @dvdherron. 

**Header**: Email Course Sign Up
**Subheader**: Be among the first to discover new layout possibilities with anchor positioning.

Reviewer 2: Ok, this should be fairly easy. I can get try it out now.

Reviewer 2: @SondraE @jamesnw 

![image](https://github.com/user-attachments/assets/f1ce6094-e6f2-4763-a9fc-1c62a2b37eb1)

Reviewer 3: lgtm!

Reviewer 2: I realized the section heading was in the wrong column and adjusted that in https://github.com/oddbird/oddleventy/pull/826/commits/ba5768356048413b89ded9f14519bbbb57b0dfc9

![Uploading image.png…]()
